Enable running phpcs-changed locally against a branch

coreymckrill opened the pull request:
### Changes proposed in this Pull Request:

Following up on #37465, this adds a way to run phpcs-changed against commits in a branch, rather than against staged or unstaged changes in the working directory. The way that phpcs-changed works, it needs to have another branch ("base" branch) to compare to for determining which lines have changed and thus should be linted. This script assumes the base branch is trunk, but another branch name can optionally be specified when running the command.

This also updates the other lint commands to use double brackets rather than single for testing the contents of the `$chg` variable, to ensure they work correctly when multiple changed files are involved (see discussion in this PR).

### How to test the changes in this Pull Request:

1. Check out this branch so that you have the new linting script available.
2. Check out another, new branch based off of this one. E.g. `git checkout -b test-lint-branch`
3. On this new branch, make a change to a line that has an existing linting error in a file with lots of existing linting errors. [Here's a good candidate.](https://github.com/woocommerce/woocommerce/blob/c4fc54680b79f306234ba43c0c82f186f643cf1b/plugins/woocommerce/templates/cart/cart.php#L80)
4. Commit that change.
5. Now in your terminal, from the repo root, change to `plugins/woocommerce`.
6. Run the new command: `composer run lint-branch`. The linter should identify errors with the line you changed in your commit, but not other lines in the file.

coreymckrill committed:
Add lint-branch.sh

coreymckrill committed:
Add doc block and make var name more explicit

coreymckrill committed:
Add changelog file

jorgeatorres commented:
Hi @coreymckrill,

I tried the steps you outlined but I'm getting an error:

```
> sh ./bin/lint-branch.sh
./bin/lint-branch.sh: line 17: [: includes/class-wc-cart.php: binary operator expected
```

The same error affects the other `lint` commands, so I think it's because `$chg` (and in this particular case `$changedFiles`) is not expanding correctly or something.

I then tried to replicate the steps from the the `bin/lint-branch.sh` script and the final `composer exec` fails with this:

```
phpcs-changed: An error occurred.
ERROR: Cannot read file 'includes/class-wc-cart.php
templates/cart/cart.php'
Script phpcs-changed handling the __exec_command event returned with error code 1
```

This makes me think the issue _is_ in fact related to expansion of the `$changedFiles` var, but I'm not exactly sure what's going on. FWIW, I'm using `zsh` as shell.

Could you take a look when you have the chance? It might very well be that I'm doing something wrong 😅.

coreymckrill commented:
@jorgeatorres 🤔 My first thought here is that it might be because you still have an older version of `phpcs-changed`, before [the fix](https://github.com/woocommerce/woocommerce/issues/36850#issuecomment-1483841136) that enabled these commands. Try going to `plugins/woocommerce/bin/composer/phpcs` and running `composer i` from there, and see if that changes your results...

jorgeatorres commented on plugins/woocommerce/bin/lint-branch.sh:
```diff
@@ -0,0 +1,17 @@
+changedFiles=$(git diff $(git merge-base HEAD $baseBranch) --relative --name-only -- '*.php')
+
+# Only complete this if changed files are detected.
+[ -z $changedFiles ] || composer exec phpcs-changed -- -s --git --git-base $baseBranch $changedFiles
```
Enclosing `$changedFiles` in double quotes allows the script to work ok when multiple files have been modified. See https://github.com/woocommerce/woocommerce/pull/38085#issuecomment-1547913636.

coreymckrill replied:
Oops, from your other comment I got fixated on the brackets thing and missed that this was about quotes 🤦

jorgeatorres reviewed (changes requested):
Requesting just a minor change, per my comment above.

coreymckrill committed:
Update composer's lint and lint-staged scripts

It turns out that using single brackets for a test doesn't work
sometimes when the variable contains spaces or multiple items.
Double brackets handles this better, so should make the scripts more
robust.

coreymckrill committed:
Use double brackets in lint-branch script

Using double brackets for a test is more robust and may be able to
handle some cases that single brackets can't, such as a variable that
contains spaces or multiple items.

Note that this also required updating the shebang at the top of the
script file from `sh` to `bash` since double brackets is a syntax
introduced in Bash.

coreymckrill commented:
@jorgeatorres I wasn't able to replicate your issue (I'm also using zsh), but I switched to double brackets anyway. Does that fix the issue for you?

jorgeatorres commented:
Hey @coreymckrill,

I wasn't suggesting the use of double brackets but double quotes, though for the conditional check that seems to do the same in the end as `[[` supports multi-word while `[` doesn't.

I believe most shells should support that, but the explicit use of `bash` probably has us covered. As such, I'm approving. We can later revisit if the use of an apparently Bash-specific extension proves problematic.

FWIW, I tested with multiple files (which was the scenario giving me trouble before) and it works fine now.

Thanks! 💯

jorgeatorres reviewed (approved):
LGTM.
